dev/dev-env-setup.sh

```diff
@@ -0,0 +1,254 @@
+      PYENV_ROOT="$HOME/.pyenv"
+      PYENV_BIN="$PYENV_ROOT/bin"
+      if [ "$MLFLOW_DEV_ENV_CI_RUN" == 1 ]; then
+        echo "$PYENV_BIN" >> "$GITHUB_PATH"
+        echo "PYENV_ROOT=$PYENV_ROOT" >> "$GITHUB_ENV"
+      fi
+    elif [[ "$machine" == win ]]; then
+      if [ -z "$(command -v pip)" ]; then
+        echo "A pip installation cannot be found. Install pip first."
+        exit 1
+      fi
+      # install via system pip as per pyenv-win docs
+      pip install $(quietpip) pyenv-win --target "$HOME\\.pyenv"
+      if [ "$MLFLOW_DEV_ENV_CI_RUN" == 1 ]; then
+        {
+          echo "PYENV=$USERPROFILE\.pyenv\pyenv-win\\";
+          echo "PYENV_ROOT=$USERPROFILE\.pyenv\pyenv-win\\";
+          echo "PYENV_HOME=$USERPROFILE\.pyenv\pyenv-win\\";
+        } >> "$GITHUB_ENV"
+        echo "$USERPROFILE\.pyenv\pyenv-win\\bin\\" >> "$GITHUB_PATH"
+      fi
+    else
+      echo "Unknown operating system environment: $machine exiting."
+      exit 1
+    fi
+  else
+    if [[ "$machine" == win ]]; then
+      PYENV_README=https://github.com/pyenv-win/pyenv-win/blob/master/README.md
+    elif [[ "$machine" == mac || "$machine" == linux ]]; then
+      PYENV_README=https://github.com/pyenv/pyenv/blob/master/README.md
+    else
+      echo "The current OS is unknown. Please visit: https://github.com/pyenv/pyenv#installation for instructions. "
+      exit 1
+    fi
+    echo "pyenv is required to use this environment setup script. Please install by following instructions here: $PYENV_README"
+    exit 1
+  fi
+fi
+
+MLFLOW_HOME=$(pwd)
+rd="$MLFLOW_HOME/requirements"
+
+# Get the minimum supported version from MLflow to ensure any feature development adheres to legacy Python versions
+min_py_version=$(grep "python_requires=" "$MLFLOW_HOME/setup.py" | grep -E -o "([0-9]{1,}\.)+[0-9]{1,}")
+
+echo "The minimum version of Python to ensure backwards compatibility for MLflow development is: $(tput bold; tput setaf 3)$min_py_version$(tput sgr0)"
+
+echo "The top-level dependencies that will be installed are: "
+
+if [[ $full_install == 1 ]]; then
+  files=("$rd/dev-requirements.txt")
+  echo "Files:"
+  echo "MLflow test plugin: $MLFLOW_HOME/tests/resources/mlflow-test-plugin"
+  echo "The local development branch of MLflow installed in editable mode with 'extras' requirements"
+  echo "All dependencies of the top-level requirements listed in the following files located in $rd: "
+else
+  files=("$rd/small-requirements.txt" "$rd/lint-requirements.txt" "$rd/large-requirements.txt" "$rd/doc-requirements.txt")
+fi
+echo "$(tail -n +1 $files | grep "^[^#= ]")"
+
+# Resolve a minor version to the latest micro version
+case $min_py_version in
+  "3.7") PY_INSTALL_VERSION="3.7.13" ;;
+  "3.8") PY_INSTALL_VERSION="3.8.13" ;;
+  "3.9") PY_INSTALL_VERSION="3.9.11" ;;
+  "3.10") PY_INSTALL_VERSION="3.10.3" ;;
+esac
+
+# Install the Python version if it cannot be found
+pyenv install -s $PY_INSTALL_VERSION
```

Comment:
I suggest use `pyenv install -s -v $PY_INSTALL_VERSION` 
the `-v` make pyenv print compilation status to stdout which helps debugging  when error happens.

and shall we add a "pyenv_root_dir" argument to allow user to specify where pyenv download the python tarballs to ?

Reply:
> the -v make pyenv print compilation status to stdout which helps debugging when error happens.

Have you tried running `pyenv install -v`? It's too verbose?

Reply:
> shall we add a "pyenv_root_dir" argument to allow user to specify where pyenv download the python tarballs to ?

I don't see a need for that argument . In what situation would that argument be useful?

Reply:
> Have you tried running pyenv install -v? It's too verbose
Yes it might be too verbose but it helps provides more clues when compiling failed. 

> I don't see a need for that argument . In what situation would that argument be useful?
E.g the case the system disk remaining capacity is not sufficient, we can specify the path to be on other disk. Just a minor suggestion.

Reply:
See https://github.com/pyenv/pyenv/blob/fab0082bd5cdda07f0bfdd69a9c676bc2d2906b3/plugins/python-build/bin/python-build#L135. pyenv stores build logs in a temporary file.

Reply:
> E.g the case the system disk remaining capacity is not sufficient, we can specify the path to be on other disk. Just a minor suggestion.

Can we just remove some files to free disk space?